syntax = "proto3";

package faiss;

import "google/protobuf/empty.proto";

message Vector {
    repeated float float_val = 5; 
}

message SearchRequest {
    Vector vector = 1;
    uint64 top_k = 2;
};

message Neighbor {
    uint64 id = 1;
    float score = 2;
}

message SearchResponse {
    repeated Neighbor neighbors = 2;
};

message SearchByIdRequest {
    uint64 id = 1;
    uint64 top_k = 2;
};

message SearchByIdResponse {
    uint64 request_id = 1;
    repeated Neighbor neighbors = 2;
};

service FaissService {
    rpc Search(SearchRequest) returns (SearchResponse);
    rpc SearchById(SearchByIdRequest) returns (SearchByIdResponse);
};
